test/core/xds/xds_matcher_e2e_test.cc

```diff
@@ -0,0 +1,650 @@
+    if (upb_proto == nullptr) {
+      EXPECT_TRUE(false) << "upb parsing failed";
+      return nullptr;
+    }
+    return upb_proto;
+  }
+
+  // Helper to parse the proto and return a validated matcher list
+  std::unique_ptr<XdsMatcher> ParseMatcherProto(const std::string& text_proto) {
+    xds::type::matcher::v3::Matcher matcher_proto;
+    EXPECT_TRUE(google::protobuf::TextFormat::ParseFromString(text_proto,
+                                                              &matcher_proto));
+
+    const auto* m_upb = ConvertToUpb(matcher_proto);
+    EXPECT_NE(m_upb, nullptr) << "Failed to convert to UPB";
+    if (m_upb == nullptr) return nullptr;
+
+    ValidationErrors errors;
+    auto context_name = RpcMatchContext::Type();
+    // Create registry and register the test-only factory.
+    XdsMatcherActionRegistry action_registry;
+    action_registry.AddActionFactory(std::make_unique<StringActionFactory>());
+    ValidationErrors::ScopedField field(&errors, ".matcher");
+    auto matcher = ParseXdsMatcher(decode_context_, m_upb, action_registry,
+                                   context_name, &errors);
+    EXPECT_TRUE(errors.ok()) << errors.status(
+        absl::StatusCode::kInvalidArgument, "unexpected errors");
+    EXPECT_NE(matcher, nullptr);
+    return matcher;
+  }
+
+  // Helper function to parse a matcher proto and check for expected errors.
+  void ParseMatcherProtoAndExpectError(const std::string& text_proto,
+                                       const std::string& expected_error) {
+    xds::type::matcher::v3::Matcher matcher_proto;
+    ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(text_proto,
+                                                              &matcher_proto));
+    const auto* m_upb = ConvertToUpb(matcher_proto);
+    ASSERT_NE(m_upb, nullptr);
+
+    ValidationErrors errors;
+    auto context_name = RpcMatchContext::Type();
+    XdsMatcherActionRegistry action_registry;
+    action_registry.AddActionFactory(std::make_unique<StringActionFactory>());
+    ValidationErrors::ScopedField field(&errors, ".matcher");
+    auto matcher = ParseXdsMatcher(decode_context_, m_upb, action_registry,
+                                   context_name, &errors);
+    absl::Status status = errors.status(absl::StatusCode::kInvalidArgument,
+                                        "Matcher parsing failed");
+    EXPECT_EQ(status.code(), absl::StatusCode::kInvalidArgument);
+    EXPECT_EQ(status.message(),
+              absl::StrCat("Matcher parsing failed: [", expected_error, "]"))
+        << status;
+  }
+
+  // Helper to append metadata safely
+  void AppendMetadata(grpc_metadata_batch& batch, absl::string_view key,
+                      absl::string_view value) {
+    batch.Append(key, Slice::FromCopiedString(value),
+                 [](absl::string_view, const Slice&) { abort(); });
+  }
+
+  // Helper to verify a match result
+  void VerifyMatchResult(XdsMatcher* matcher, grpc_metadata_batch& metadata,
```

Comment:
Given that we already have tests in xds_matcher_test.cc that verify the actual matcher behavior, it seems redundant to test that again here.  Instead, I suggest having these tests just check that the parsing code constructed the proper `XdsMatcher` object by calling the `ToString()` method.